linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.java

```diff
@@ -0,0 +1,228 @@
+
+  @Autowired private EngineConnLaunchService engineConnLaunchService;
+
+  @Autowired private LabelFilter labelFilter;
+
+  @Autowired private ECResourceInfoService ecResourceInfoService;
+
+  @Autowired private LabelResourceService labelResourceService;
+
+  @Override
+  public List listEngines(GetEMEnginesRequest getEMEnginesRequest) {
+    AMEMNode emNode = new AMEMNode();
+    emNode.setServiceInstance(getEMEnginesRequest.getEm());
+    return emNodeManager.listEngines(emNode);
+  }
+
+  @Override
+  public EngineNode createEngine(EngineConnBuildRequest engineBuildRequest, EMNode emNode) {
+    logger.info(
+        "EM " + emNode.getServiceInstance() + " start to create Engine " + engineBuildRequest);
+    EngineConnLaunchRequest engineConnLaunchRequest =
+        engineConnLaunchService.createEngineConnLaunchRequest(engineBuildRequest);
+
+    EngineNode engineNode = emNodeManager.createEngine(engineConnLaunchRequest, emNode);
+    logger.info(
+        "EM "
+            + emNode.getServiceInstance()
+            + " Finished to create Engine "
+            + engineBuildRequest.ticketId());
+    engineNode.setLabels(
+        emNode.getLabels().stream()
+            .filter(label -> label instanceof EngineNodeLabel)
+            .collect(Collectors.toList()));
+    engineNode.setEMNode(emNode);
+    return engineNode;
+  }
+
+  @Override
+  public void stopEngine(EngineNode engineNode, EMNode emNode) {
+    if (emNode == null) {
+      logger.error("The emNode of " + engineNode.getServiceInstance() + " is null");
+      return;
+    }
+    logger.debug(
+        "EM "
+            + emNode.getServiceInstance()
+            + " start to stop Engine "
+            + engineNode.getServiceInstance());
+    EngineStopRequest engineStopRequest = new EngineStopRequest();
+    engineStopRequest.setServiceInstance(engineNode.getServiceInstance());
+
+    engineStopRequest.setIdentifierType(engineNode.getMark());
+    engineStopRequest.setIdentifier(engineNode.getIdentifier());
+
+    ECResourceInfoRecord ecResourceInfo =
+        ecResourceInfoService.getECResourceInfoRecordByInstance(
+            engineNode.getServiceInstance().getInstance());
+    if (ecResourceInfo != null) {
+      engineStopRequest.setEngineType(ecResourceInfo.getLabelValue().split(",")[1].split("-")[0]);
+      engineStopRequest.setLogDirSuffix(ecResourceInfo.getLogDirSuffix());
+    } else {
+      if (CollectionUtils.isEmpty(engineNode.getLabels())) {
+        // node labels is empty, engine already been stopped
+        logger.info(
+            "DefaultEMEngineService stopEngine node labels is empty, engine: {} have already been stopped.",
+            engineStopRequest.getServiceInstance());
+        return;
+      }
+
+      RMLabelContainer rMLabelContainer = labelResourceService.enrichLabels(engineNode.getLabels());
+
+      PersistenceResource persistenceResource =
+          labelResourceService.getPersistenceResource(rMLabelContainer.getEngineInstanceLabel());
```

Comment:
Because ecResourceInfo is already been null in above line, so that I get Ticket id from  labelResourceService ,please correct me if I was wrong ,thanks.
![image](https://user-images.githubusercontent.com/16526279/236677569-e4c3d1d1-31ec-42b6-b06d-913b57a38c61.png)

Reply:
Sorry, I misread